File: Algorithm/220915/SWEA_노드의 합.cpp
#include <iostream>
//#include <fstream>
using namespace std;

int t, n, m, l, in_a, in_b, i, head;
int tree[1001];

int postorder(int cur) {
	if (cur > n)
		return 0;
	else {
		if (tree[cur] == 0)
			tree[cur] = postorder(cur * 2) + postorder(cur * 2 + 1);
		return tree[cur];
	}
}

int main() {
	//ifstream cin("input.txt");
	cin >> t;
	for (int cs = 0; cs < t; cs++) {
		cin >> n >> m >> l;
		for (i = 1; i <= n; i++)
			tree[i] = 0;
		for (i = 0; i < m; i++) {
			cin >> in_a >> in_b;
			tree[in_a] = in_b;
		}
		head = postorder(1);
		cout << "#" << cs + 1 << " " << tree[l] << endl;
	}
	//cin.close();
	return 0;
}

File: Algorithm/220915/SWEA_노드의 합.py
import sys
sys.stdin = open("input.txt")

def postorder(cur):                                             # 재귀함수로 풀어냄
    if cur > n:
        return 0                                                # 그래프 범위 밖으로 벗어났으면 0을 리턴 (덧셈에 지장 X)
    elif lst[cur] != 0:
        return lst[cur]                                         # 리프 노드거나 이미 값이 기록됐다면 그 값을 리턴
    else:
        lst[cur] = postorder(cur * 2) + postorder(cur * 2 + 1)  # 부모 노드에 자식 노드의 합을 기록
        return lst[cur]                                         # 부모 노드의 값을 리턴

t = int(input())
for cs in range(t):
    n, m, l = map(int, input().split())
    lst = [0] * (n + 1)
    for i in range(m):
        in_a, in_b = map(int, input().split())
        lst[in_a] = in_b

    postorder(1)
    print(f'#{cs + 1} {lst[l]}')